File: examples/GoogleSearchTest.java
/**
 * Google search test for Java / JUnit 5 — Appium drop-in against drizz-farm.
 *
 * Drive Chrome mobile via UiAutomator2, search, assert results.
 * drizz:* capabilities opt in to video / logcat capture on the daemon side.
 *
 * Maven dependencies (appium-java-client pulls in Selenium + HTTP client):
 *   <dependency>
 *     <groupId>io.appium</groupId>
 *     <artifactId>java-client</artifactId>
 *     <version>9.3.0</version>
 *   </dependency>
 *   <dependency>
 *     <groupId>org.junit.jupiter</groupId>
 *     <artifactId>junit-jupiter</artifactId>
 *     <version>5.11.0</version>
 *     <scope>test</scope>
 *   </dependency>
 *
 * Run:
 *   DRIZZ_URL=http://parthas-macbook-pro.local:9401 mvn test
 */
import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.android.options.UiAutomator2Options;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.net.URI;
import java.time.Duration;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertFalse;

public class GoogleSearchTest {

    private static final String DRIZZ_URL =
        System.getenv().getOrDefault("DRIZZ_URL", "http://localhost:9401");
    private static final String QUERY =
        System.getenv().getOrDefault("QUERY", "drizz farm open source device lab");

    private AndroidDriver driver;

    @BeforeEach
    void setUp() throws Exception {
        UiAutomator2Options opts = new UiAutomator2Options()
            .setPlatformName("Android")
            .setAutomationName("UiAutomator2")
            .setBrowserName("Chrome")
            // drizz-farm extras — /wd/hub strips these out and starts
            // captures before forwarding cleaned caps to Appium.
            .amend("drizz:record_video",       true)
            .amend("drizz:capture_logcat",     true)
            .amend("drizz:capture_screenshots", true)
            .amend("drizz:timeout_minutes",    10);

        driver = new AndroidDriver(URI.create(DRIZZ_URL + "/wd/hub").toURL(), opts);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(5));
        System.out.println("[drizz] session " + driver.getSessionId());
    }

    @Test
    void searchForDrizzFarm() {
        driver.get("https://www.google.com/ncr");
        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(15));

        // Consent screen (EU / fresh emulators) — tap accept if visible.
        try {
            WebElement accept = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//button[.//span[contains(., 'Accept') or contains(., 'agree')]]")));
            accept.click();
            System.out.println("[drizz] consent accepted");
        } catch (Exception ignored) {
        }

        WebElement box = wait.until(ExpectedConditions.elementToBeClickable(By.name("q")));
        box.clear();
        box.sendKeys(QUERY);
        box.submit();

        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("search")));

        List<WebElement> results = driver.findElements(By.cssSelector("#search a h3"));
        System.out.println("[drizz] got " + results.size() + " results");
        for (int i = 0; i < Math.min(5, results.size()); i++) {
            System.out.println("  " + (i + 1) + ". " + results.get(i).getText());
        }
        assertFalse(results.isEmpty(), "expected at least one search result");
    }

    @AfterEach
    void tearDown() {
        if (driver != null) {
            String sid = driver.getSessionId().toString();
            driver.quit();
            System.out.println("[drizz] artifacts: " + DRIZZ_URL + "/api/v1/sessions/" + sid + "/artifacts");
            System.out.println("[drizz] playback : " + DRIZZ_URL + "/playback/" + sid);
        }
    }
}

File: examples/google_search_test.py
"""
Google search test against drizz-farm's Appium-compat /wd/hub endpoint.

Point DRIZZ_URL at your farm (default http://localhost:9401). This does
NOT install an app — it drives Chrome via Appium's UiAutomator2, opens
google.com, types a query, taps search, asserts results are visible.
Because it uses no installed APK, it works on any Play-image emulator.

drizz:* capabilities opt in to video + logcat + HAR capture. After the
test, fetch artifacts from /api/v1/sessions/<session_id>/artifacts.

Run:
    pip install appium-python-client
    DRIZZ_URL=http://parthas-macbook-pro.local:9401 python examples/google_search_test.py
"""
import os
import time

from appium import webdriver
from appium.options.android import UiAutomator2Options
from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait


DRIZZ_URL = os.environ.get("DRIZZ_URL", "http://localhost:9401")
QUERY = os.environ.get("QUERY", "drizz farm open source device lab")


def main() -> None:
    opts = UiAutomator2Options()
    opts.platform_name = "Android"
    opts.automation_name = "UiAutomator2"
    # Drive the built-in Chrome instead of a custom APK so the test
    # works on any Play-store emulator image.
    opts.browser_name = "Chrome"
    # drizz-farm extras — the Appium server ignores unknown caps; our
    # /wd/hub compat layer peels them off and starts the captures
    # before forwarding the cleaned caps to Appium.
    opts.set_capability("drizz:record_video", True)
    opts.set_capability("drizz:capture_logcat", True)
    opts.set_capability("drizz:capture_screenshots", True)
    opts.set_capability("drizz:timeout_minutes", 10)

    hub = f"{DRIZZ_URL}/wd/hub"
    print(f"[drizz] connecting to {hub}")
    driver = webdriver.Remote(hub, options=opts)
    session_id = driver.session_id
    print(f"[drizz] session {session_id}")

    try:
        driver.get("https://www.google.com/ncr")  # ncr = no-country-redirect, stable UI

        wait = WebDriverWait(driver, 15)
        # Handle the consent screen if it shows (EU / new install).
        try:
            accept = wait.until(EC.element_to_be_clickable(
                (By.XPATH, "//button[.//span[contains(., 'Accept') or contains(., 'agree')]]")
            ))
            accept.click()
            print("[drizz] consent accepted")
        except Exception:
            pass

        search_box = wait.until(EC.element_to_be_clickable(
            (By.NAME, "q")
        ))
        search_box.clear()
        search_box.send_keys(QUERY)
        search_box.submit()
        print(f"[drizz] searched: {QUERY}")

        # Wait for the search results container.
        wait.until(EC.presence_of_element_located((By.ID, "search")))
        time.sleep(2)  # let results settle for the video

        results = driver.find_elements(By.CSS_SELECTOR, "#search a h3")
        print(f"[drizz] got {len(results)} results")
        for i, r in enumerate(results[:5], 1):
            print(f"  {i}. {r.text}")
        assert len(results) > 0, "expected at least one search result"
        print("[drizz] PASS")
    finally:
        driver.quit()

    # Drizz-farm finalizes video.mp4 / logcat.txt on session release.
    print(f"[drizz] artifacts: {DRIZZ_URL}/api/v1/sessions/{session_id}/artifacts")
    print(f"[drizz] playback : {DRIZZ_URL}/playback/{session_id}")


if __name__ == "__main__":
    main()
